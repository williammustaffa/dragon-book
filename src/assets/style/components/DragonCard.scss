.dragon-card {
  .card-image {
    height: 200px;
    width: 100%;
    background-size: cover;
    background-position: center;
    background-color: #2e2e2e;

    .card-actions {
      transition: 0.2s ease-in-out;
      height: 100%;
      width: 100%;
      display: flex;
      align-items: center;
      justify-content: center;
      background-color: rgba(0, 0, 0, 0.5);
      opacity: 0;
      color: $color-white;
      font-size: $font-size-big;
    }

    .card-action {
      margin: $spacing;
      transform-origin: 50% 50%;

      &:hover {
        transform: scale(1.1);
      }
    }

    &:hover {
      .card-actions {
        opacity: 1;
      }
    }
  }
}