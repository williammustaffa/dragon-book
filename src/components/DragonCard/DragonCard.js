import React from "react";
import PropTypes from "prop-types";
import HTMLEllipsis from "react-lines-ellipsis/lib/html"
import { push } from "connected-react-router";
import { useDispatch } from "react-redux";
import { Card, Label } from "semantic-ui-react";

function DragonCard({ dragon }) {
  const dispatch = useDispatch();
  const navigateTo = path => () => dispatch(push(path));

  return (
    <Card fluid className="dragon-card" onClick={navigateTo(dragon.url)}>
      <div
        className="card-image"
        style={{ backgroundImage: `url(${dragon.imageUrl})` }}
      ></div>
      <Card.Content>
        <Card.Header onClick={navigateTo(dragon.url)}>
          <span >{dragon.name}</span>
        </Card.Header>
        <Card.Meta>
          Created at <span className="link static">{dragon.creationDate}</span>
        </Card.Meta>
        <Card.Description>
          <HTMLEllipsis
            unsafeHTML={dragon.shortDescription}
            maxLine="3"
            ellipsis="..."
            basedOn="letters"
          />
        </Card.Description>
      </Card.Content>
      <Card.Content extra>
        <Label
          className="text-right text-uppercase"
          attached="bottom"
          color={dragon.typeColor}
          size="small"
        >
          {dragon.type}
        </Label>
      </Card.Content>
    </Card>
  );
}

DragonCard.propTypes = {
  dragon: PropTypes.object.isRequired,
};

export default DragonCard;