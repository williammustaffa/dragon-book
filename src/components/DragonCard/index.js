export { default as DragonCard } from "./DragonCard";