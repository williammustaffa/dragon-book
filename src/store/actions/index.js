// Session
export { default as userLogin } from "./user/userLogin";
export { default as userLoginSuccess } from "./user/userLoginSuccess";
export { default as userLoginFailure } from "./user/userLoginFailure";
export { default as userLogout } from "./user/userLogout";
export { default as userCheckSession } from "./user/userCheckSession";
export { default as userCheckSessionSuccess } from "./user/userCheckSessionSuccess";
export { default as userCheckSessionFailure } from "./user/userCheckSessionFailure";

// Profile
export { default as createProfile } from "./profiles/createProfile";
export { default as createProfileFailure } from "./profiles/createProfileFailure";
