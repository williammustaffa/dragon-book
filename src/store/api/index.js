
import database from "store/api/database";
import * as R from "ramda";
import { v1 as uuidv1 } from 'uuid';
import request from "./request";

const AUTH_TOKEN_KEY = "auth.token";

class API {
  /**
   * Fake API error delay
   * @param {String} value 
   * @param {Int} time 
   */
  static retrieveDataDelayed(value, time = 500) {
    return new Promise(resolve => {
      setTimeout(() => resolve(value), time);
    });
  }

  /**
   * Fake API error delay
   * @param {String} value 
   * @param {Int} time 
   */
  static retrieveErrorDelayed(value, time = 500) {
    return new Promise((resolve, reject) => {
      setTimeout(() => reject({ message: value }), time);
    });
  }

  /**
   * Create token for user
   * @param {*} param
   */
  static tokenize({ id, email }) {
    return btoa(JSON.stringify({ id, email }));
  }

  /**
   * Get user info from token
   */
  static getTokenizedInfo() {
    const token = localStorage.getItem(AUTH_TOKEN_KEY);
  
    return token && JSON.parse(atob(token));
  }

  /**
   * Parse response object
   * @param {*} res 
   */
  static parseReponse(res) {
    return res.data;
  }

  /**
   * Create a user profile
   */
  createProfile = (payload) => {
    const errors = [];

    const existingUserEmail = database.queryAll("profiles", {
      query: { email: payload.email }
    });

    const existingUserUsername = database.queryAll("profiles", {
      query: { username: payload.username }
    });

    if (existingUserEmail.length) {
      errors.push("The email address provided is already in use.");
    }

    if (existingUserUsername.length) {
      errors.push("The username provided is already in use.");
    }

    if (errors.length) {
      return API.retrieveErrorDelayed(errors.join("\r\n"));
    }

    database.insert("profiles", {
      id: uuidv1(),
      email: payload.email,
      username: payload.username,
      firstName: payload.firstName,
      lastName: payload.lastName,
      description: payload.description,
      password: payload.password,
      registrationDate: Date.now(),
    });

    database.commit();

    // Create user and login
    return this.userLogin(payload);
  }

  /**
   * Check user session an then populate data if needed
   */
  userCheckSession = () => {
    const data = API.getTokenizedInfo();

    const results = database.queryAll("profiles", {
      query: { id: data && data.id },
    });

    if (!results.length) {
      return API.retrieveErrorDelayed("No session found.");
    }

    const user = R.clone(results[0]);

    // No need to delay on first load
    return user;
  }

  /**
   * Do user login by given crendentials
   * @param {object} credentials 
   */
  userLogin({ email, password }) {
    const results = database.queryAll("profiles", {
      query: { email, password },
    });

    if (!results.length) {
      return API.retrieveErrorDelayed("You have entered an invalid username or password");
    }

    const user = R.clone(results[0]);

    // Fake a token storage
    localStorage.setItem(AUTH_TOKEN_KEY, API.tokenize(user));

    return API.retrieveDataDelayed(user, 3000);
  }

  /**
   * Clear user session when he loogs out
   */
  userLogout = () => {
    // Clear token
    localStorage.removeItem(AUTH_TOKEN_KEY);

    return API.retrieveDataDelayed(null, 0);
  }

  /**
   * List dragons
   */
  listDragons = () => {
    return request
      .get("/dragon")
      .then(API.parseReponse);
  }

  /**
   * Get dragon details
   * @param {String} id
   */
  getDragon = (id) => {
    return request
      .get(`/dragon/${id}`)
      .then(API.parseReponse);
  }

  /**
   * Get dragon details
   * @param {Object} data
   */
  createDragon = (data) => {
    return request
      .post("/dragon", data)
      .then(API.parseReponse);
  }

  /**
   * Update dragon details
   * @param {Object} data
   */
  updateDragon = (data) => {
    return request
      .put(`/dragon/${data.id}`, data)
      .then(API.parseReponse);
  }

  /**
   * Update dragon details
   * @param {String} data
   */
  deleteDragon = (id) => {
    return request
      .put(`/dragon/${id}`)
      .then(API.parseReponse);
  }
}

// Create a cached connector
let connectorInstance = null;

function getAPIConnector() {
  if (!connectorInstance) {
    connectorInstance = new API();
  }

  return connectorInstance;
}


export { API, getAPIConnector };