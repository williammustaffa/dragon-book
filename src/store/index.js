import { applyMiddleware, compose, createStore } from "redux";
import { routerMiddleware } from "connected-react-router";
import createRootReducer from "./reducers";
import sagas from "./sagas";
import history from "./history";
import createSagaMiddleware from "redux-saga"
import logger from "redux-logger";
import configureStore from "redux-mock-store";

/**
 * Generates a redux store properly
 * applying reducers, states and middlewares
 *
 * @param {*} preloadedState 
 */
function generateStore(preloadedState) {
  const sagaMiddleware = createSagaMiddleware();

  const store = createStore(
    createRootReducer(history),
    preloadedState,
    compose(
      applyMiddleware(
        routerMiddleware(history),
        sagaMiddleware,
        logger,
      ),
    ),
  );

  // Run redux saga
  sagaMiddleware.run(sagas);

  return store;
}

/**
 * Mock store and return a mocked state
 * @param {*} preloadedState 
 */
function mockStore(preloadedState) {
  const middlewares = [];
  const generateMockStore = configureStore(middlewares);

  return generateMockStore(preloadedState);
}

export {
  generateStore,
  mockStore,
  history,
};