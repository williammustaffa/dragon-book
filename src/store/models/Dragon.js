import moment from "moment";

class Dragon {
  constructor(payload = {}) {
    this.id = payload.id || "";
    this.createdAt = payload.createdAt;
    this.name = payload.name || "";
    this.type = payload.type || "";
    this.histories = payload.histories || [];
    this.imageUrl = payload.imageUrl;
  }

  /**
   * Returns the dragon detail url
   */
  get url() {
    return `/dragon/${this.id}`;
  }

  /**
   * Returns a formatted creation date
   * @returns {String} creation date
   */
  get creationDate() {
    return (
      this.createdAt &&
      moment(new Date(this.createdAt)).format("MMM Do YY")
    );
  }

  /**
   * Returns a color based on the dragon type
   * @returns {String} color
   */
  get typeColor() {
    const colors = {
      legendary: "yellow",
      epic: "purple",
      rare: "blue",
      normal: "grey",
      default: "black"
    };

    return colors[this.type] || colors.default;
  }

  /**
   * Returns a computed short desciption
   * @returns {String} description
   */
  get shortDescription() {
    return this.histories
      .join("\n")
      .substring(0, 200);
  }

  /**
   * Initial reducer state
   */
  static state = {
    isFetching: false,
    details: new Dragon(),
    errorMessage: ""
  }
}

export default Dragon;