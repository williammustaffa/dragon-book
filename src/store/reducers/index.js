import { combineReducers } from "redux";
import { connectRouter } from "connected-react-router";
import dragons from "./dragons";
import dragon from "./dragon";
import user from "./user";

const createRootReducer = (history) => combineReducers({
  router: connectRouter(history),
  dragons,
  dragon,
  user,
});

export default createRootReducer;