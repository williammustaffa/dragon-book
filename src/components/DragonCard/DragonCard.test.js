import React from "react";
import Dragon from "store/models/Dragon";
import { Provider } from 'react-redux';
import { mockStore } from "store";
import DragonCard from "./DragonCard";
import { render } from "@testing-library/react";
import "@testing-library/jest-dom/extend-expect";

test("Test dragon card component", () => {
  // Mock store
  const store = mockStore({});

  // Create mock dragon object
  const dragon = new Dragon({
    id: "1",
    name: "Saphira",
    type: "Epic",
    createdAt: "Sun May 10 2020 11:10:53 GMT-0300",
    histories: [],
  });

  const { asFragment } = render(
    <Provider store={store}>
      <DragonCard dragon={dragon}/>
    </Provider>
  );

  // Run tests
  expect(asFragment()).toMatchSnapshot();
});