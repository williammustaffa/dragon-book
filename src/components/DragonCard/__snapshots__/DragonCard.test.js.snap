// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test dragon card component 1`] = `
<DocumentFragment>
  <a
    class="ui fluid card dragon-card"
  >
    <div
      class="clickable card-image"
      style="background-image: url(undefined);"
    />
    <div
      class="content"
    >
      <div
        class="header"
      >
        <span
          class="clickable"
        >
          Saphira
        </span>
      </div>
      <div
        class="meta"
      >
        Created at 
        <span
          class="link static"
        >
          May 10th 20
        </span>
      </div>
      <div
        class="description clickable"
      >
        <div
          class="LinesEllipsis  "
        >
          <div />
        </div>
      </div>
    </div>
    <div
      class="extra content dragon-type"
    >
      <div
        class="ui black small bottom attached label"
      >
        Epic
      </div>
    </div>
  </a>
</DocumentFragment>
`;
